### ISSUE TITLE
Multi-inheritance `super(SuperA, obj)` doesn't propery return the next `mro()` super

### ISSUE THREAD

**Issue by Kache** [microsoft/pyright#6660]
**Describe the bug**

`super()` has a cooperative multiple inheritance use-case that traverses `mro()`:

https://docs.python.org/3.11/library/functions.html?highlight=classmethod#super

However, pyright doesn't seem to return the correct superclass. Below, `b` is not correctly identified as a `B`.

**Code or Screenshots**
[pyright playground](https://pyright-play.net/?pyrightVersion=1.1.338&pythonVersion=3.11&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqAmgAoCiA%2BgMIASZVA0gJIByA4gFAcDGANgIYBnQVACCALg5RpUACYBTYFGBgwACkHzewAJSSZBqCHkwAriBRQA5KKtc%2BQkQCF9MhUpXrN2vVMPTjMwtrJzseAWEoKjVRABooJ18DBEcuHigAXii1HQ5%2BTKhBUwR5EDUqeO5cgCMCopKyuKgqrgQQVBg1eBK1KoA6CHAcnWkAYigAbQAeB0irCgoIflQFvqorAD54mYiRecXllFXbLagdx2sFpZWKPtDT87mwaoAreW4YTYBdVvaUTv4fU8Iyg4ym1TAphQsigEBMAAswDDREDVFAwEopldDqsqOjXu98Px8AAGAAeAEYSQAWABsAGZ5AAmWQkjYbDhtDpqQGeYZjMQcJBKYjkah0RisNiuALyABu8n4vAo3XkPJB4wIiHk6KUACJ%2BHrMCI9aI9b9udVUWANWcIVCYXCYIiYU5rbqztibmt8W8PlBiVByVS6YyWWyOVz-morXydLanEKRaRKLR6Mx2DKjPLFcrVTHbVqSh69dUjUgTc8-TA9UA)

```python
from typing import TYPE_CHECKING

class A:
    def foo(self):
        return 'A'

class B:
    def foo(self):
        return 'B'

class C(A, B):
    pass


c = C()
a = super(C, c)
b = super(A, c)

print(type(c).mro())  # [<class '__main__.C'>, <class '__main__.A'>, <class '__main__.B'>, <class 'object'>]

print(a.foo)  # <bound method A.foo of <__main__.C object at 0x10463e2d0>>
print(a.foo())  # A
if TYPE_CHECKING:
    reveal_type(a)  # Type of "a" is "A"

print(b.foo)  # <bound method B.foo of <__main__.C object at 0x10463e2d0>>
print(b.foo())  # B
if TYPE_CHECKING:
    reveal_type(b)  # Type of "b" is "object"
```

**VS Code extension or command-line**
CLI `pyright 1.1.338`

When searching issues I found possibly related #151

**Comment by erictraut** [comment]
In general, the type of `super()` cannot be determined statically unless it's used within an attribute access expression. At runtime, `super` returns a proxy object that performs a dynamic lookup of the accessed attribute. Without knowing which attribute is being accessed, it's not possible to predict which class will be dynamically chosen.

To make matters worse, the MRO for a class cannot always be statically determined. For example, if a class is used as a mixin for another class, it can affect the MRO for the resulting class.

If you use the `super` call with an attribute access expression, pyright is able to statically determine the right answer in most cases.

```python
c = C()
a = super(C, c).foo()
reveal_type(a)  # Literal['A']

b = super(A, c).foo()
reveal_type(b)  # Literal['B']
```

If you use `super` without an attribute access expression, as you're doing in your example above, there are limits to what a static type checker can do. Mypy evaluates the type of `super(C, c)` as `builtins.super` and doesn't allow any attribute to be subsequently accessed from it. I don't think that's the right behavior.

I think it's reasonable to employ a heuristic here — one that chooses the next base class of the bound type. This won't be accurate 100% of the time, but it will work for common use cases. I've updated pyright's logic to use this heuristic.

This behavior may change in the future based on input or clarification in the typing spec, but for now this should meet your needs.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
